### ISSUE TITLE
SPtam crashing during runtime

### ISSUE THREAD

**Issue by siddyyagno3** [lrse/sptam#10]
Configuration:
OS: Ubuntu 14.04
ROS distro: Indigo
Camera:Logitech c270
Camera Driver: uvc_camera
Laptop processor: corei7

I was trying to run sptam on my webcam pair to get odometry data.
While running the launch file I get this warning recursively . There is no window opened till I move the camera.

```
sidd@sidd-K55VM:~/catkin_ws$ roslaunch sptam lau.launch
... logging to /home/sidd/.ros/log/b87c8c84-f9c9-11e5-bb92-94dbc9b19dba/roslaunch-sidd-K55VM-10100.log
Checking log directory for disk usage. This may take awhile.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://sidd-K55VM:44428/

SUMMARY
========

PARAMETERS
 * /left_decimator/height: 860
 * /left_decimator/width: 1177
 * /left_decimator/x_offset: 52
 * /left_decimator/y_offset: 51
 * /right_decimator/height: 860
 * /right_decimator/width: 1177
 * /right_decimator/x_offset: 52
 * /right_decimator/y_offset: 51
 * /rosdistro: indigo
 * /rosversion: 1.11.16
 * /sptam/DescriptorExtractor/BytesLength: 32
 * /sptam/DescriptorExtractor/Name: BRIEF
 * /sptam/DescriptorMatcher/Name: BruteForce-Hamming
 * /sptam/DescriptorMatcher/crossCheck: False
 * /sptam/EpipolarDistance: 0
 * /sptam/FeatureDetector/Name: GFTT
 * /sptam/FeatureDetector/minDistance: 15.0
 * /sptam/FeatureDetector/nfeatures: 1000
 * /sptam/FeatureDetector/qualityLevel: 0.01
 * /sptam/FeatureDetector/useHarrisDetector: False
 * /sptam/FramesBetweenKeyFrames: 0
 * /sptam/FrustumFarPlaneDist: 10000
 * /sptam/FrustumNearPlaneDist: 0.1
 * /sptam/KeyFrameDistance: 0
 * /sptam/MatchingCellSize: 30
 * /sptam/MatchingDistance: 25
 * /sptam/MatchingNeighborhood: 1
 * /sptam/approximate_sync: True
 * /sptam/baseline: 0.071
 * /sptam/camera_frame: camera_1
 * /sptam/camera_matrix: cv::Mat
 * /sptam/cols: 3
 * /sptam/data: [718.856, 0, 607....
 * /sptam/dt: d
 * /sptam/image_height: 480
 * /sptam/image_width: 640
 * /sptam/odom_frame: odom_combined
 * /sptam/rows: 3
 * /sptam/use_odometry: False

NODES
/stereo/
    stereo_image_proc (stereo_image_proc/stereo_image_proc)
  /
    camera_broadcaster (tf/static_transform_publisher)
    left_decimator (nodelet/nodelet)
    right_decimator (nodelet/nodelet)
    sptam (sptam/sptam_node)

ROS_MASTER_URI=http://localhost:11311

core service [/rosout] found
process[stereo/stereo_image_proc-1]: started with pid [10118]
process[left_decimator-2]: started with pid [10119]
process[right_decimator-3]: started with pid [10120]
process[camera_broadcaster-4]: started with pid [10121]
process[sptam-5]: started with pid [10138]
[ INFO] [1459708372.712452032]: S-PTAM node running...
detector: GFTT
  minDistance: 15
  nfeatures: 1000
  qualityLevel: 0.01
  useHarrisDetector: 0
extractor: BRIEF
matcher: BruteForce-Hamming
  crossCheck: 0
[ INFO] [1459708373.133859507]: sptam node initialized
[ INFO] [1459708374.761461761]: dt: 0
[ INFO] [1459708374.761563392]: init calib
[ INFO] [1459708374.761623424]: tx: -368.944 fx: 805.413 baseline: -0.458081
[ INFO] [1459708374.761892478]: baseline: 0.386787
[ INFO] [1459708374.787497848]: Trying to intialize map...
No matches found for triangulation
[ INFO] [1459708374.820623114]: dt: 0
[ INFO] [1459708374.838212296]: Trying to intialize map...
No matches found for triangulation
[ INFO] [1459708374.891819060]: dt: 0
[ INFO] [1459708374.913047833]: Trying to intialize map...
No matches found for triangulation
.
.
.
```

Not sure if my calibration files are right.

> Distance between 2 camera is 7cm. I tried to enter it in calibration file but I think it doesn't load it.
> The following command was added in launch file.
>  <rosparam command="load" file="$(find sptam)/configurationFiles/cam.yaml" />

[cam_yaml.txt](https://github.com/lrse/sptam/files/201494/cam_yaml.txt)

[my_yaml.txt](https://github.com/lrse/sptam/files/201495/my_yaml.txt)

This is the launch file vaguely based on level7.launch
[lau_launch.txt](https://github.com/lrse/sptam/files/201496/lau_launch.txt)

When I move my camera during runtime it opens a window and start identifying and tracking features. And node data can be visualized in rviz by subscribing to pose. Not sure if it was proper movement of pose as it moves the pointer somewhat according to my motion. 
Terminal shows the following.

```
[ INFO] [1459711365.654362880]: Trying to intialize map...
[ INFO] [1459711365.696182493]: dt: 0
[ INFO] [1459711365.730848287]: Trying to intialize map...
Points initialized from stereo: 11
[ INFO] [1459711365.760036317]: dt: 0
init done 
opengl support available 


WARNING: Not enough points for tracking.

 Adding key-frame.
1.45973e+09 tk trackingtotal: 0.235
Se agrega/n 1 KeyFrame al mapa.
BA Local: kf adj: 1 kf fix: 1 points: 7
EJECUTANDO BA LOCAL:

[ INFO] [1459711365.995880619]: dt: 0
DESPUES DE BA LOCAL

BA LOCAL CONVERGIO
mediciones borradas por el BA: 4


WARNING: Not enough points for tracking.
```

This window is opened while camera is in motion and crashes when I stop the motion.
![screenshot from 2016-04-04 00 44 01](https://cloud.githubusercontent.com/assets/9764434/14234326/63624600-f9fe-11e5-84d7-4336cd875ffe.png)

The error on terminal when it crashes

```
BA LOCAL CONVERGIO
sptam_node: /home/sidd/catkin_ws/src/sptam/src/sptam/thirdparty/MEstimator.hpp:79: static double Tukey::FindSigmaSquared(std::vector<double>&): Assertion `vdErrorSquared.size() > 0' failed.
[ INFO] [1459711530.416988060]: dt: 0
[sptam-5] process has died [pid 19897, exit code -6, cmd /home/sidd/catkin_ws/devel/lib/sptam/sptam_node /robot/pose:=/odom __name:=sptam __log:=/home/sidd/.ros/log/b87c8c84-f9c9-11e5-bb92-94dbc9b19dba/sptam-5.log].
log file: /home/sidd/.ros/log/b87c8c84-f9c9-11e5-bb92-94dbc9b19dba/sptam-5*.log

```

Have to press ctrl+c to end the process.
This is the log file.
[roslaunch_log.txt](https://github.com/lrse/sptam/files/201512/roslaunch_log.txt)

This is the error in log file.

> [roslaunch.pmon][INFO] 2016-04-04 00:42:02,694: registrations completed <ProcessMonitor(ProcessMonitor-1, started daemon 140200667428608)>
> [roslaunch.parent][INFO] 2016-04-04 00:42:02,694: ... roslaunch parent running, waiting for process exit
> [roslaunch][INFO] 2016-04-04 00:42:02,694: spin
> [roslaunch][ERROR] 2016-04-04 00:42:42,132: [sptam-5] process has died [pid 11786, exit code -6, cmd /home/sidd/catkin_ws/devel/lib/sptam/sptam_node /robot/pose:=/odom __name:=sptam __log:=/home/sidd/.ros/log/b87c8c84-f9c9-11e5-bb92-94dbc9b19dba/sptam-5.log].
> log file: /home/sidd/.ros/log/b87c8c84-f9c9-11e5-bb92-94dbc9b19dba/sptam-5*.log
> [roslaunch.pmon][INFO] 2016-04-04 00:42:42,133: ProcessMonitor.unregister[sptam-5] starting

 I can't seem to figure out why I am getting this error. 
Thanks in advance.

**Comment by taihup** [comment]
As you say, S-PTAM can not crate an initial map. This could be caused by a wrong camera calibration, or the scene does not have enough features. Try to recalibrate the camera with [http://wiki.ros.org/camera_calibration/Tutorials/StereoCalibration](url) or with [http://www.vision.caltech.edu/bouguetj/calib_doc/](url).

On the other hand, take care that you are copping the image in the launch file as level7.launch file. Try to crop the image propperly or not cropping it at all.

Let me know, if this helps?

Taihú

**Comment by siddyyagno3** [comment]
Thank you for your reply. 

We have calibrated the cameras and are currently not cropping. S-PTAM is working and we can see the point clouds in RVIZ. We have another couple problems as a follow up.
1. When we set _use_odometry to "true" by running in launchfile as well as by using the command:<br>
   `rosrun sptam sptam_node _approximate_sync:=true _use_odometry:=true` <br>
   the topics <br>
   `/S_PTAM/point_cloud` <br>
   `/S_PTAM/robot/pose` <br>

```

NODES
  /stereo/
    stereo_image_proc (stereo_image_proc/stereo_image_proc)
  /
    camera_broadcaster (tf/static_transform_publisher)
    sptam (sptam/sptam_node)
    sptam_path (ros_utils/pose_to_path)

ROS_MASTER_URI=http://localhost:11311

core service [/rosout] found
process[camera_broadcaster-1]: started with pid [20084]
ERROR: cannot launch node of type [ros_utils/pose_to_path]: ros_utils
ROS path [0]=/opt/ros/indigo/share/ros
ROS path [1]=/home/sidd/catkin_ws/src
ROS path [2]=/opt/ros/indigo/share
ROS path [3]=/opt/ros/indigo/stacks
process[stereo/stereo_image_proc-3]: started with pid [20085]
process[sptam-4]: started with pid [20086]
[ INFO] [1459782485.377107995]: S-PTAM node running...
detector: GFTT
  minDistance: 15
  nfeatures: 1000
  qualityLevel: 0.01
  useHarrisDetector: 0
extractor: BRIEF
matcher: BruteForce-Hamming
  crossCheck: 0
[ INFO] [1459782485.791753899]: sptam node initialized
[ INFO] [1459782487.473742022]: dt: 0
[ WARN] [1459782487.574682077]: Failed to retrieve camera pose in odom frame
[ INFO] [1459782487.575199268]: dt: 0
[ WARN] [1459782487.676182246]: Failed to retrieve camera pose in odom frame
[ INFO] [1459782487.676580956]: dt: 0
[ WARN] [1459782487.777517507]: Failed to retrieve camera pose in odom frame
[ INFO] [1459782487.777988269]: dt: 0

```

do not publish anything.
It works when set to false. 
1. Is there any way to access odometry data to visualize on rviz? Through some topic that can be published or some other method? Or have you created a rviz config file which you can share with us? 
2. In the cam.yaml file is there any parameters we can vary such as the <br>
   `data: [ 718.856, 0, 607.1928, 0, 718.856, 185.2157, 0, 0, 1 ]` <br>
   parameter or camera matrix  for better running? 

Thanks again.

**Comment by taihup** [comment]
Hi siddyyagno3, are you running the S-PTAM in a robot (or dataset) which provides odometry ROS messages? If not, you can not set true the use_odometry flag.
1. I think you can subscribe to odometry messages with RVIZ without problem.
2. What cam.yaml are you talking about? Looks you are talking about camara calibration parameters (intrinsic matrix K). However, this parameters are provided by cameraInfo ROS messages to S-PTAM.

Best!

**Comment by siddyyagno3** [comment]
Sorry for losing track of this issue. 
Got it working thanks to taihup .
After reinstalling and bit of luck I realized that installation of g2o ROS package was wrong and had to install the recommended one: svn co https://svn.openslam.org/data/svn/g2o .
And the distance between camera for good output in indoor environment was above 10cm.
Implemented it on our bot it was working and used it follow a predefined path.
We are working towards using this odometry package combined with path following to drive any general bot and publish it on git.
